c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1529,6 +1529,12 @@ class DeviceCachingAllocator {
         EventPool::Event event = std::move(e.first);
         Block* block = e.second;
 
+        const c10::impl::PyInterpreter* interp =
+            c10::impl::GPUTrace::get_trace();
+        if (C10_UNLIKELY(interp)) {
+          (*interp)->trace_gpu_event_synchronization(
+              reinterpret_cast<uintptr_t>(*event));
+        }
```

Comment:
This is happening within the caching allocator, but I thought that CSAN aimed to "sit above" the allocator? So shouldn't we avoid adding a hook here and just treat the sanitizer as a safe black box?

Moreover, if I understand correctly, this code is run when the sanitizer needs to free memory, and this could happen non-deterministically (i.e., depending on memory pressure, and when the PyTorch garbage allocator decides to delete the tensors). Thus if this event sync happens "at the right time" it might make by accident it look like the user's code is safe, whereas the user's code was actually unsafe if the sync happened at another time.

Reply:
You're right, this can indeed nondeterministically cause synchronizations that result in a false positive, and the sanitizer should never depend on this as a synchronization. I will remove the hook.